syntax = "proto3";

enum Message {
    MESSAGE_HANDSHAKE = 0;
    MESSAGE_REQUEST = 1;
    MESSAGE_RESPONSE = 2;
    MESSAGE_HEARTBEAT = 3;
}

message Handshake {
    uint32 timeout = 1;
    uint32 outgoing_window_size = 2;
    uint32 incoming_window_size = 3;
    bytes id = 4;
}

message RequestHeader {
    uint32 sequence_number = 1;
    bytes service_name = 2;
    uint32 method_index = 3;
}

message Request {
    RequestHeader header = 1;
    bytes payload = 2;
}

enum Error {
    ERROR_NO = 0;
    ERROR_CHANNEL_BUSY = 1;
    ERROR_NOT_IMPLEMENTED = 2;
    ERROR_BAD_REQUEST = 3;
    ERROR_INTERNAL_SERVER = 4;
    ERROR_USER_DEFINED = 256;
}

message ResponseHeader {
    uint32 sequence_number = 1;
    uint32 error_code = 2;
}

message Response {
    ResponseHeader header = 1;
    bytes payload = 2;
}

message Heartbeat {
}
